Add geometry conformance

twhiteaker opened the pull request:
Add conformance rules for geometries. See trac ticket 164 and CF conventions pull request 115 for geometries discussion.

twhiteaker committed:
Fix grammar in introduction

twhiteaker committed:
add geomery requirements

twhiteaker committed:
Clarify that *nodes* for polygon rings must be in a certain order.

twhiteaker committed:
Remove case sensitive requirement for geometry type, and remove recommendation of first-last node equivalence for polygons

twhiteaker committed:
require axis on coordinate vars

twhiteaker committed:
Merge branch 'Geometry_Conformance'

* Geometry_Conformance:
  require axis on coordinate vars
  Remove case sensitive requirement for geometry type, and remove recommendation of first-last node equivalence for polygons
  Clarify that *nodes* for polygon rings must be in a certain order.
  add geomery requirements

twhiteaker committed:
Revise geometry conformance per David Hassell request

davidhassell commented on conformance.adoc:
```diff
@@ -552,6 +552,8 @@ a geometry container variable. The variable name must exist in the file.
 * The only legal values of geometry_type are **`point`**, **`line`**,
 and **`polygon`** (case insensitive).
+* For a line **`geometry_type`**, each geometry must have a minimum of two node coordinates.
+* For a polygon **`geometry_type`**, each geometry must have a minimum of three node coordinates.
```
Shouldn't "three" be "two", in which case, we could merge two rules into one:

  * For a line or polygon `geometry_type`, each geometry must have a minimum of two node coordinates.

E.g. In Example 7.16. (Polygons with holes), we have only two node coordinates. (https://github.com/twhiteaker/cf-conventions/blob/master/ch07.adoc)

twhiteaker replied:
I was referring to the size of the node dimension, not the number of node coordinate variables.  Is this rule clearer?

* For a line **`geometry_type`**, each geometry must have a minimum of two nodes.

davidhassell replied:
Ah - I see - my mistake! I think the original text is fine. I had confused "node coordinates" with "node coordinate variables". it is clear in chapter 7, so I think that it is OK here, too.

davidhassell commented on conformance.adoc:
```diff
@@ -561,29 +563,25 @@ particular value of the **`axis`** attribute.
 geometries, and within each multipart geometry the nodes must be stored
 consecutively for each part and in the order of the parts.
 * Nodes for polygon exterior rings must be put in anticlockwise order (viewed from above)
 and polygon interior rings in clockwise order.
```
The above two rules concerning the order of nodes are not checkable, so shouldn't they be omitted from the conformance requirements? These points are already made in chapter 7. ​

twhiteaker replied:
> * The nodes must be stored consecutively for each geometry...

That one is uncheckable.  I agree with omitting it.

> * Nodes for polygon exterior rings must be put in anticlockwise order...

This one can be checked.  [Here's a checker in Python](https://github.com/twhiteaker/netCDF-CF-simple-geometry/blob/master/src/python/ncgeom/part.py#L98).

davidhassell replied:
That's good for me. Thanks for the link to your geometry code.

twhiteaker replied:
I removed the uncheckable rule.  You can find the latest version here:
https://github.com/cf-convention/Conformance/pull/6/files

twhiteaker committed:
Remove uncheckable rule about nodes being in consecutive order per geometry and part

davidhassell commented on conformance.adoc:
```diff
@@ -568,9 +568,6 @@ variable, then those coordinate variables which correspond to node coordinate va
  **`bounds`** attribute that names the corresponding node coordinate.
```
Following the conclusion of the "nodes" discussion, this check needs updating. How about:

* If a coordinate variable named by a **`coordinates`** attribute carried by the geometry container variable or its parent data variable has a **`nodes`** attribute, then the **`nodes`** attribute must be a string whose value is a single variable name. The specified variable must be a node coordinate variable that exists in the file.

twhiteaker committed:
change bounds to nodes for coordinate variables and geometry

twhiteaker committed:
Allow alternate grid mappings for geometry containers

Fixes #156

twhiteaker committed:
require coordinates to use same grid mapping as nodes

twhiteaker committed:
Remove geometry_dimension per cf conventions issue #155

When node_count attribute is missing, require the dimension of the node coordinate variables to be one of the dimensions of the data variable.

dblodgett-usgs commented:
Given that https://github.com/cf-convention/cf-conventions/pull/158 has been merged. I'm going to go ahead and merge this too.
